src/cache-utils.ts

```diff
@@ -66,37 +107,191 @@ export const getPackageManagerInfo = async (packageManager: string) => {
+ * @return list of files or directories expanded from glob
+ */
+const globPatternToArray = async (pattern: string): Promise<string[]> => {
+  const memoized = expandedPatternsMemoized[pattern];
+  if (memoized) return Promise.resolve(memoized);
+  const globber = await glob.create(pattern);
+  const expanded = await globber.glob();
+  expandedPatternsMemoized[pattern] = expanded;
+  return expanded;
+};
 
-    if (yarnVersion.startsWith('1.')) {
-      return supportedPackageManagers.yarn1;
-    } else {
-      return supportedPackageManagers.yarn2;
-    }
+/**
+ * Expands (converts) the string input `cache-dependency-path` to list of files' paths
+ * First it breaks the input by new lines and then expand glob patterns if any
+ * @param cacheDependencyPath - either a single string or multiline string with possible glob patterns
+ * @return list of files on which the cache depends
+ */
+export const expandCacheDependencyPath = async (
+  cacheDependencyPath: string
+): Promise<string[]> => {
+  const multilinePaths = cacheDependencyPath
+    .split(/\r?\n/)
+    .map(path => path.trim())
+    .filter(path => Boolean(path));
+  const expandedPathsPromises: Promise<string[]>[] = multilinePaths.map(path =>
+    path.includes('*') ? globPatternToArray(path) : Promise.resolve([path])
+  );
+  const expandedPaths: string[][] = await Promise.all(expandedPathsPromises);
+  return expandedPaths.length === 0 ? [''] : expandedPaths.flat();
+};
+
+/**
+ * Converts dependency file path to the directory it resides in and ensures the directory exists
+ * @param cacheDependencyPath - a file name path
+ * @return either directory containing the file or null
+ */
+const cacheDependencyPathToProjectDirectory = (
+  cacheDependencyPath: string
+): string | null => {
+  const projectDirectory = path.dirname(cacheDependencyPath);
+  if (
+    fs.existsSync(projectDirectory) &&
+    fs.lstatSync(projectDirectory).isDirectory()
+  ) {
+    core.debug(
+      `Project directory "${projectDirectory}" derived from cache-dependency-path: "${cacheDependencyPath}"`
+    );
+    return projectDirectory;
   } else {
+    core.debug(
+      `No project directory found for cache-dependency-path: "${cacheDependencyPath}", will be skipped`
+    );
     return null;
   }
 };
 
-export const getCacheDirectoryPath = async (
-  packageManagerInfo: PackageManagerInfo,
-  packageManager: string
-) => {
-  const stdOut = await getCommandOutput(
-    packageManagerInfo.getCacheFolderCommand
+/**
+ * Expands (converts) the string input `cache-dependency-path` to list of directories that
+ * may be project roots
+ * @param cacheDependencyPath - either a single string or multiline string with possible glob patterns
+ *                              expected to be the result of `core.getInput('cache-dependency-path')`
+ * @return list of directories and possible
+ */
+const cacheDependencyPathToProjectsDirectories = async (
+  cacheDependencyPath: string
+): Promise<string[]> => {
```

Comment:
```suggestion
const getProjectDirectories = async (
  cacheDependencyPath: string
): Promise<string[]> => {
```

Reply:
I'd completely remove function `expandCacheDependencyPath()` because I believe it's not needed. Why? Because function `getInput()` always returns a string. And this string will be well formatted to be supplied to the `glob.create()` right away.
**Example:**
```yaml
- uses: actions/setup-node@v3
   with:
      cache-dependency-path: |
         path1
         path2
         path3
```
If we check the `cacheDependencyPath` input data it will look like: "path1\npath2\npath3".

So I'd use `glob` package right away to find all files matched by the `cacheDependecyPath`. Next I'd convert the list of paths to lock files, to list of paths to directories which contain these lock files. Then, to make this array unique, I'd convert it to Set and then return back Array made from Set using Array.from().

Reply:
The only point is using Set implies memory allocation for binary tree that is too expensive for the small number of elements, the direct search is used instead